### PR TITLE
chore: add helper utility functions for foundation deploy tasks

### PR DESCRIPTION
**Pull request by daniel-cit**
This PR adds helper utility functions for future foundation automate deploy tasks

### CODE REVIEW

File: helpers/deployer/utils/git.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	b, err := conf.RunCmdE("branch", "-q", "--show-current")
+	if err != nil {
+		return "", err
+	}
+	return b, nil
+}
+
+// HasRemoteTRacking check it a branch has a remote upstream configured.
+func HasRemoteTRacking(conf *git.CmdCfg, branch string) (bool, error) {
+	s, err := conf.RunCmdE("status", "-sb")
```
Comment (bharathkkb) [inline helpers/deployer/utils/git.go:48]: Couldn't we `git remote get-url origin` to validate a remote called origin?

File: helpers/deployer/utils/git.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+		_, err := conf.RunCmdE("push", "--set-upstream", "origin", branch)
+		if err != nil {
+			return err
+		}
+	} else {
+		_, err := conf.RunCmdE("push")
+		if err != nil {
+			return err
+		}
+	}
```
Comment (bharathkkb) [inline helpers/deployer/utils/git.go:75]: The `"--set-upstream` is a no-op if it's already set right?

File: helpers/deployer/utils/git.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+// CheckoutBranch checkouts a branch.
+// If the branch does not exist it will be created.
+func CheckoutBranch(conf *git.CmdCfg, branch string) error {
+	c, err := GetCurrentBranch(conf)
+	if err != nil {
+		return err
+	}
+	if c != branch {
```
Comment (bharathkkb) [inline helpers/deployer/utils/git.go:86]: nit: we can switch this to return early
```
if c == branch {
return nil
}
...
```

File: helpers/deployer/utils/git.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		_, err := conf.RunCmdE("checkout", "-b", branch)
+		if err != nil {
+			if strings.Contains(err.Error(), "already exists") {
```
Comment (bharathkkb) [inline helpers/deployer/utils/git.go:89]:
```suggestion
		if err != nil && strings.Contains(err.Error(), "already exists")
```

File: helpers/deployer/utils/tfvars.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+// ReadTfvars reads a valid terraform tfvars file into the provided struct.
+func ReadTfvars(filename string, val interface{}) error {
+	f, d := hclparse.NewParser().ParseHCLFile(filename)
+	if d.HasErrors() {
+		return errors.New(d.Error())
+	}
+	d = gohcl.DecodeBody(f.Body, nil, val)
+	if d.HasErrors() {
+		return errors.New(d.Error())
```
Comment (bharathkkb) [inline helpers/deployer/utils/tfvars.go:34]: IIRC you can just return d (diag) as it satisfies the error iface.

**Review by bharathkkb — COMMENTED** [review]
Thanks @daniel-cit! Some initial comments

File: helpers/deployer/utils/git.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	"github.com/GoogleCloudPlatform/cloud-foundation-toolkit/infra/blueprint-test/pkg/git"
+	"github.com/gruntwork-io/terratest/modules/logger"
+	"github.com/mitchellh/go-testing-interface"
+)
+
+// CloneRepo clones a Google cloud source repository and returns a CmdConfig pointing to the repository.
+func CloneRepo(t testing.TB, name, path, project string) *git.CmdCfg {
+	_, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		gcloud.Runf(t, "source repos clone %s %s --project %s", name, path, project)
```
[the remark this answers is not in the document]
Comment (daniel-cit) [reply helpers/deployer/utils/git.go:32]: We don't need to configure the helper manually.
When the repo is cloned using gcloud, gcloud sets this configuration locally 
```
 credential.https://source.developers.google.com/.helper=!gcloud auth git-helper --account=<USER-EAMIL> --ignore-unknown $@
```

File: helpers/deployer/utils/files.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+		return err
+	}
+	return os.WriteFile(filename, bytes.Replace(f, []byte(old), []byte(new), -1), 0644)
+}
+
+// FindFiles find files with the given filename under the directory.
+func FindFiles(dir, filename string) ([]string, error) {
+	found := []string{}
+	err := filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
+		if d.Name() == filename && !strings.Contains(path, ".terraform") {
```
[the remark this answers is not in the document]
Comment (daniel-cit) [reply helpers/deployer/utils/files.go]: fixed

File: helpers/deployer/utils/files_test.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestCopyFile(t *testing.T) {
+	d, err := os.Getwd()
+	assert.NoError(t, err)
+	src := filepath.Join(d, "files_test.go")
```
[the remark this answers is not in the document]
Comment (daniel-cit) [reply helpers/deployer/utils/files_test.go]: fixed

File: helpers/deployer/utils/git.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+	}
+	return git.NewCmdConfig(t, git.WithDir(path), git.WithLogger(logger.Default))
+}
+
+// GetCurrentBranch getes the current branch in the repository.
```
[the remark this answers is not in the document]
Comment (daniel-cit) [reply helpers/deployer/utils/git.go]: fixed

File: helpers/deployer/utils/git.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+func GetCurrentBranch(conf *git.CmdCfg) (string, error) {
+	b, err := conf.RunCmdE("branch", "-q", "--show-current")
+	if err != nil {
+		return "", err
+	}
+	return b, nil
+}
```
[the remark this answers is not in the document]
Comment (daniel-cit) [reply helpers/deployer/utils/git.go]: fixed

File: helpers/deployer/utils/git.go (hunk 2 above)
Comment (daniel-cit) [reply helpers/deployer/utils/git.go]: yes, it does not return an error

File: helpers/deployer/utils/git.go (hunk 3 above)
Comment (daniel-cit) [reply helpers/deployer/utils/git.go]: fixed

File: helpers/deployer/utils/git.go (hunk 4 above)
Comment (daniel-cit) [reply helpers/deployer/utils/git.go]: fixed

File: helpers/deployer/utils/tfvars.go (hunk 5 above)
Comment (daniel-cit) [reply helpers/deployer/utils/tfvars.go]: good catch

**Review by bharathkkb — APPROVED** [review]

**Comment by shark696969** [comment]
is:aproved resolved

Pada tanggal Sel, 4 Apr 2023 06.49, Bharath KKB ***@***.***>
menulis:

**Pull request merged**